### ISSUE TITLE
Await and Awilix Cradle

### ISSUE THREAD

**Issue by rclayton-the-terrible** [jeffijoe/awilix#109]
Hi!  We ran into a small issue today and I was curious how you felt about it.

We are abstracting out DI providers from our platform, having used Inverisfy in the past and wanting to give our teams options to choose the provider they want.  Personally, I like Awilix and have been adding a module for my own project.

One assumption we made about providers is that they would conform to the following interface:

```typescript
import { Dependencies } from '../api/Dependencies'

import Context from '../api/Context'
import Subscriber from '../api/Subscriber'
import Handler from '../api/Handler'

export default interface DependenciesProvider {
    load(context: Context<any>, handler: Handler | Subscriber): Promise<Dependencies>
    unload(dependencies: Dependencies, context: Context<any>, handler: Handler | Subscriber): Promise<void>
}
```

You can ignore the function parameters and `Dependencies` (which is basically `{ [k:string]: any }`).  The Awilix implementation is very simple:

```typescript
import { Dependencies } from '../../api/Dependencies'
import { AwilixContainer } from 'awilix'

import DependenciesProvider from '../DependenciesProvider'
import Context from '../../api/Context'
import Handler from '../../api/Handler'
import Subscriber from '../../api/Subscriber'

export default class AwilixDependenciesProvider implements DependenciesProvider {

    constructor(protected container: AwilixContainer) {}

    async load(context: Context<any>, handler: Handler | Subscriber): Promise<Dependencies> {
        return this.container.createScope().cradle
    }

    async unload(dependencies: Dependencies, context: Context<any>, handler: Handler | Subscriber): Promise<void> {}
}
```

The interface requires `async` loads because we want to allow some async actions in the future.

This leads to a problem, which is particularly difficult with TypeScript compilations:

```typescript
describe('when pulling deps from the provider', () => {
        it('should return the requested dependencies from the proxy', async () => {
            const provider = new AwilixDependenciesProvider(container)
            const { time, foo, theAnswer } = await provider.load(
                new MockContext(envelope.data, envelope),
                {} as Handler
            )
            expect(time).toBeInstanceOf(Time)
            expect(foo).toEqual('bar')
            expect(theAnswer).toEqual(42)
        })
    })
```

Is resulting in:

```
AwilixResolutionError: Could not resolve 'then'.

    Resolution path: then
```

Since the `cradle` is a proxy and resolving property access, `await` is causing it to try to resolve `then` when it's not actually a `Promise`.

Clearly, this is not Awilix's fault; but in many settings, it's hard to predict when compiled languages like TypeScript will unnecessarily await a non-Promise.

We have solved it with this ugly piece of code:

```typescript
container.register({
  then: asValue(undefined)
})
```

I think the question becomes, is this something worth adding to the framework?

**Comment by jeffijoe** [comment]
How about `return Promise.resolve(cradle)`?

**Comment by jeffijoe** [comment]
Is this does not work, I would suggest either do as you're doing there by having `then` be registered, or by wrapping it in your own proxy. 😄

**Comment by andyfleming** [comment]
The following solves the issue:

```
container.register({
      then: asValue(undefined),
```

In my opinion, the container should disallow registering of `then` as a key and should automatically resolve it to `undefined` (as the default behavior).

This would make it async-await and promise friendly.

**Comment by rclayton-the-terrible** [comment]
@jeffijoe because of the weirdness around how TypeScript compiles code, returning `Promise.resolve(cradle)` did not work for us.  We ended up having the `cradle` awaited as well.

Also, another proxy wrapping the original proxy feels like a matryoshka doll.

**Comment by jeffijoe** [comment]
To be honest I don't think you should be using async await when resolving dependencies and therefore I don't think it is a good idea to add that because it would imply that it is a good idea.

If you insist on doing it anyway, I think the trick you posted should do the trick. 😀

**Comment by rclayton-the-terrible** [comment]
Well, hopefully, this thread will help someone who runs into a similar problem.  @jeffijoe thank you for the quick responses.

**Comment by jeffijoe** [comment]
Excellent, no problem!

**Comment by andyfleming** [comment]
Thanks for the quick replies @jeffijoe.

The use case isn't at the point of resolving a dependency, but rather returning the cradle object. I do think there are valid contexts where this would be done in an async function (or similar).

It's an unexpected behavior and more the fault of javascript than the library, but it would be nice for it to elegantly handle it.

I'd be happy to write the PR if you're open to it.

**Comment by jeffijoe** [comment]
@andyfleming I think use of the cradle needs to be restricted to dereferencing as if it was an object. In my book anything that mixes async with the cradle is a smell.

But alright, feel free to submit a PR with the suggested changes. Internally there are multiple proxies (to support injectors) so make sure those are covered by tests too. 😀

**Comment by jeffijoe** [comment]
@andyfleming @rclayton-the-terrible a fix has been released in v4.0.1 - thanks for the report and @andyfleming for the PR! 🎉
